src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java

```diff
@@ -0,0 +1,148 @@
+ * The stream supplier should be repeatable, as the entity data might be produced multiple times.
+ */
+public class StreamEntityProducer implements AsyncEntityProducer {
+
+    private static final Logger logger = org.slf4j.LoggerFactory.getLogger(StreamEntityProducer.class);
+
+    private final Supplier<InputStream> streamSupplier;
+    private final boolean chunked;
+
+    private ByteBuffer content;
+
+    private final static int BUFFER_SIZE = 8192;
+    private final byte[] buffer = new byte[BUFFER_SIZE];
+
+    private InputStream currentStream;
+
+    /**
+     * Creates a new entity producer that produces the entity data from the given input stream supplier.
+     *
+     * @param streamSupplier the input stream supplier, should be repeatable
+     * @param chunked        whether the entity data should be sent in chunks
+     */
+    public StreamEntityProducer(Supplier<InputStream> streamSupplier, boolean chunked) throws IOException {
+        this.streamSupplier = streamSupplier;
+        this.chunked = chunked;
+
+        if (!chunked) {
+            content = ByteBuffer.wrap(streamSupplier.get().readAllBytes());
+        }
+    }
+
+    @Override
+    public boolean isRepeatable() {
+        return true;
+    }
+
+    @Override
+    public void failed(Exception cause) {
+        logger.error("Failed to produce entity data", cause);
+        if (currentStream != null) {
+            try {
+                currentStream.close();
+            } catch (IOException e) {
+                logger.error("Failed to close input stream", e);
+            }
+        }
+    }
+
+    @Override
+    public boolean isChunked() {
+        return chunked;
+    }
+
+    @Override
+    public Set<String> getTrailerNames() {
+        return null;
+    }
+
+    @Override
+    public long getContentLength() {
+        // if the content length is known (non-chunked request), return it
+        if (content != null) {
+            return content.limit();
+        }
+
+        // if the content length is unknown (chunked request), return -1
+        return -1;
+    }
+
+    @Override
+    public String getContentType() {
+        return null;
+    }
+
+    @Override
+    public String getContentEncoding() {
+        return null;
+    }
+
+    @Override
+    public void releaseResources() {
+        if (currentStream != null) {
+            try {
+                currentStream.close();
+            } catch (IOException e) {
+                logger.error("Failed to close input stream", e);
+            }
+        }
+    }
+
+    @Override
+    public int available() {
+        if (content != null) {
+            return content.remaining();
+        }
+        if (currentStream != null) {
+            try {
+                return currentStream.available();
+            } catch (IOException e) {
+                logger.error("Failed to get available bytes from input stream", e);
+            }
+        }
```

Comment:
If `content` is not null, it means the whole entity data has been read into the buffer from a stream that was taken from the stream supplier and that the content will be sent non-chunked.
In this case, the remaining bytes in the buffer are returned.

Otherwise, the data is sent in chunks. If there is currently a stream open, from which the data is being read from, the available bytes from that stream are returned.

I added these as comments.